File: src/main/kotlin/com/xs0/gqlktx/types/gql/GUnionType.java
package com.xs0.gqlktx.types.gql;

import com.xs0.gqlktx.QueryException;
import com.xs0.gqlktx.schema.builder.TypeKind;
import com.xs0.gqlktx.schema.intro.GqlIntroType;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class GUnionType extends GBaseType {
    private Set<GObjectType> members;

    public GUnionType(String name) {
        super(name);
    }

    @Override
    public TypeKind getKind() {
        return TypeKind.UNION;
    }

    public void setMembers(Set<GObjectType> members) {
        this.members = members;
    }

    public Set<GObjectType> getMembers() {
        return members;
    }

    @Override
    public boolean validAsQueryFieldType() {
        return true;
    }

    @Override
    public boolean validAsArgumentType() {
        return false;
    }

    @Override
    public void coerceValue(JsonObject raw, String key, JsonObject out) throws QueryException {
        throw new QueryException("Union type " + getName() + " can't be used as a variable");
    }

    @Override
    public void coerceValue(JsonArray raw, int index, JsonArray out) throws QueryException {
        throw new QueryException("Union type " + getName() + " can't be used as a variable");
    }

    @Override
    public void toString(StringBuilder sb) {
        sb.append("union ").append(getName());

        boolean first = true;
        for (GObjectType member : members) {
            sb.append(first ? " = " : " | ");
            first = false;
            sb.append(member.getName());
        }
        sb.append("\n");
    }

    public List<GqlIntroType> getMembersForIntrospection() {
        ArrayList<GqlIntroType> res = new ArrayList<>(members.size());
        for (GObjectType posib : members)
            res.add(posib.introspector());
        return res;
    }
}

File: src/main/kotlin/com/xs0/gqlktx/types/gql/GUnionType.kt
package com.xs0.gqlktx.types.gql

import com.xs0.gqlktx.QueryException
import com.xs0.gqlktx.dom.Value
import com.xs0.gqlktx.schema.builder.TypeKind
import com.xs0.gqlktx.schema.intro.GqlIntroType

import java.util.ArrayList
import java.util.HashSet

class GUnionType(name: String, description: String?) : GBaseType(name, description) {
    lateinit var members: Set<GObjectType>
        private set

    override val kind: TypeKind
        get() = TypeKind.UNION

    override fun coerceValue(raw: Value): Value {
        throw QueryException("Union type $name can't be used as a variable")
    }

    override fun toString(sb: StringBuilder) {
        sb.append("union ").append(name)

        var first = true
        for (member in members) {
            sb.append(if (first) " = " else " | ")
            first = false
            sb.append(member.name)
        }
        sb.append("\n")
    }

    val membersForIntrospection: List<GqlIntroType> by lazy {
        val res = ArrayList<GqlIntroType>(members.size)
        for (posib in members)
            res.add(posib.introspector)
        res
    }

    fun setMembers(gqlImpls: HashSet<GObjectType>) {
        this.members = gqlImpls
    }
}
